include/tvm/script/printer/traced_object_functor.h

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef TVM_SCRIPT_PRINTER_TRACED_OBJECT_FUNCTOR_H_
+#define TVM_SCRIPT_PRINTER_TRACED_OBJECT_FUNCTOR_H_
+
+#include <tvm/node/node.h>
+#include <tvm/runtime/logging.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/script/printer/traced_object.h>
+
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace tvm {
+namespace script {
+namespace printer {
+
+namespace {
+
+namespace detail {
+/*!
+ * \brief Helper template class to extract the type of first argument of a function
+ * \tparam FType The function type.
+ */
+template <typename FType>
+struct first_arg_type_helper;
+
+template <typename R, typename ArgOne, typename... OtherArgs>
+struct first_arg_type_helper<R(ArgOne, OtherArgs...)> {
+  using T = ArgOne;
+};
+
+/*!
+ * \brief Template alias for the type of first argument of a function
+ * \tparam FType The function type.
+ *
+ * The name of public functions are in snake case to be consistent with
+ * tvm/node/functor.h
+ */
+template <typename FType>
+using first_arg_type = typename detail::first_arg_type_helper<
+    typename tvm::runtime::detail::function_signature<FType>::FType>::T;
+}  // namespace detail
+
+}  // namespace
+
+namespace dispatch_table {
+/*
+ * Functions in dispatch_table namespace is created to reduce the binary bloat
+ * from template and also hide implementation details from this header
+ */
+
+using DispatchTable = std::unordered_map<std::string, std::vector<runtime::PackedFunc>>;
+
+constexpr const char* kDefaultDispatchToken = "";
+
+const runtime::PackedFunc& GetDispatchFunction(const DispatchTable& dispatch_table,
+                                               const String& token, uint32_t type_index);
+void SetDispatchFunction(DispatchTable* dispatch_table, const String& token, uint32_t type_index,
+                         runtime::PackedFunc f);
+}  // namespace dispatch_table
```

Comment:
Curious why we are having an extra namespace? If it's not for name conflict, let's manage to reduce the namespace nesting here

Reply:
If they are used only in `TracedObjectFunctor`, do you think we could move them to its static members?

Reply:
I removed the nested namespace. We can't move them to static members of `TracedObjectFunction`, because it's a templated class.